hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/MergeOperation.java

```diff
@@ -0,0 +1,200 @@
+
+import com.hazelcast.cache.CacheNotExistsException;
+import com.hazelcast.cache.impl.CacheService;
+import com.hazelcast.cache.impl.journal.CacheEventJournal;
+import com.hazelcast.config.EventJournalConfig;
+import com.hazelcast.config.RingbufferConfig;
+import com.hazelcast.map.impl.MapService;
+import com.hazelcast.map.impl.journal.MapEventJournal;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.ringbuffer.OverflowPolicy;
+import com.hazelcast.ringbuffer.impl.RingbufferContainer;
+import com.hazelcast.ringbuffer.impl.RingbufferService;
+import com.hazelcast.spi.BackupAwareOperation;
+import com.hazelcast.spi.ObjectNamespace;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.ServiceNamespace;
+import com.hazelcast.spi.ServiceNamespaceAware;
+import com.hazelcast.spi.SplitBrainMergeEntryView;
+import com.hazelcast.spi.SplitBrainMergePolicy;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import static com.hazelcast.ringbuffer.OverflowPolicy.FAIL;
+import static com.hazelcast.ringbuffer.impl.RingbufferDataSerializerHook.F_ID;
+import static com.hazelcast.ringbuffer.impl.RingbufferDataSerializerHook.MERGE_OPERATION;
+import static com.hazelcast.ringbuffer.impl.RingbufferService.SERVICE_NAME;
+import static com.hazelcast.util.MapUtil.createHashMap;
+
+/**
+ * Merges a batch of Ringbuffer items after a split-brain was healed.
+ */
+public class MergeOperation extends Operation implements IdentifiedDataSerializable, BackupAwareOperation, ServiceNamespaceAware {
+
+    private ObjectNamespace namespace;
+    private SplitBrainMergePolicy mergePolicy;
+    private List<SplitBrainMergeEntryView<Long, Object>> mergingEntries;
+    private OverflowPolicy overflowPolicy;
+
+    private transient RingbufferConfig config;
+    private transient RingbufferContainer<Object, Object> ringbuffer;
+    private transient Map<Long, Data> valueMap;
+
+    public MergeOperation() {
+    }
+
+    public MergeOperation(ObjectNamespace namespace, SplitBrainMergePolicy mergePolicy,
+                          List<SplitBrainMergeEntryView<Long, Object>> mergingEntries, OverflowPolicy overflowPolicy) {
+        this.namespace = namespace;
+        this.mergePolicy = mergePolicy;
+        this.mergingEntries = mergingEntries;
+        this.overflowPolicy = overflowPolicy;
+    }
+
+    @Override
+    public void beforeRun() throws Exception {
+        RingbufferService service = getService();
+        config = getRingbufferConfig(service, namespace);
+        ringbuffer = service.getOrCreateContainer(getPartitionId(), namespace, config);
+    }
+
+    @Override
+    public void run() throws Exception {
+        if (overflowPolicy == FAIL) {
+            if (ringbuffer.remainingCapacity() < 1) {
+                return;
```

Comment:
Should we check the remaining capacity after each merged entry? Since this way, checking at the beginning, we can end up with more than the configured capacity (if all `mergingEntries` are added as new)

Reply:
Hmm, as I think of it, maybe that check belongs in the `merge()` method. If there is no existing entry, we have to check for the capacity. If there is an existing entry, the capacity doesn't matter, since we may just replace the value.

Reply:
The check in `MergeOperation` would anyway never apply, since the overflow policy was always set to `OVERWRITE`. I removed it completely from the operation, since it was always the same.

I changed the implementation to pass the `remainingCapacity()` to the internal `ArrayRingbuffer.merge()` method. If there is not enough capacity when a new item has to be added, it will not merge it. So we won't overwrite entries from the larger cluster anymore.